File: LovePlay-RAC/LovePlay/Class/Zone/View/ZoneListCell.m
//
//  ZoneListCell.m
//  LovePlay
//
//  Created by Yuns on 2017/2/3.
//  Copyright © 2017年 Yuns. All rights reserved.
//

#import "ZoneListCell.h"
#import "ZoneListModel.h"

@interface ZoneListCell ()
// UI
@property (nonatomic, strong) UILabel *titleTextLabel;
@property (nonatomic, strong) UILabel *detailTextLabel;
@property (nonatomic, strong) UIImageView *imageNode;
//Data
@property (nonatomic, strong) ZoneDiscussItem *discussItem;
@end

@implementation ZoneListCell

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.contentView.backgroundColor = [UIColor whiteColor];
        [self addSubnodes];
        [self mas_subViews];
    }
    return self;
}

#pragma mark - private
- (void)addSubnodes
{
    [self.contentView addSubview:self.imageNode];
    
    [self.contentView addSubview:self.titleTextLabel];
    
    [self.contentView addSubview:self.detailTextLabel];
}

- (void)mas_subViews
{
    [_imageNode mas_makeConstraints:^(MASConstraintMaker *make) {
        make.width.height.mas_equalTo(50);
        make.left.equalTo(self.contentView).offset(10);
        make.centerY.equalTo(self.contentView);
    }];
    
    [_titleTextLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(_imageNode).offset(3);
        make.left.equalTo(_imageNode.mas_right).offset(5);
        make.right.equalTo(self.contentView).offset(-10);
    }];
    
    [_detailTextLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.equalTo(_titleTextLabel);
        make.bottom.equalTo(_imageNode.mas_bottom).offset(-3);
    }];
}

#pragma mark -  public
- (void)setupDiscussItem:(ZoneDiscussItem *)discussItem
{
    _discussItem = discussItem;
    _imageNode.imageURL = [NSURL URLWithString:_discussItem.iconUrl];
    _titleTextLabel.text = _discussItem.modelName;
    _detailTextLabel.text = _discussItem.modelDesc;
}

#pragma mark - setter / getter
- (UIImageView *)imageNode
{
    if (!_imageNode) {
        UIImageView *imageNode = [[UIImageView alloc] init];
        _imageNode = imageNode;
    }
    return _imageNode;
}

- (UILabel *)titleTextLabel
{
    if (!_titleTextLabel) {
        UILabel *titleTextLabel = [[UILabel alloc] init];
        titleTextLabel.font = [UIFont systemFontOfSize:14];
        titleTextLabel.textColor = RGB(36, 36, 36);
        _titleTextLabel = titleTextLabel;
    }
    return _titleTextLabel;
}

- (UILabel *)detailTextLabel
{
    if (!_detailTextLabel) {
        UILabel *detailTextLabel = [[UILabel alloc] init];
        detailTextLabel.font = [UIFont systemFontOfSize:12];
        detailTextLabel.textColor = RGB(150, 150, 150);
        _detailTextLabel = detailTextLabel;
    }
    return _detailTextLabel;
}
@end

File: LovePlay-Swift3/LovePlay/Class/Zone/View/ZoneListCell.swift
//
//  ZoneListCell.swift
//  LovePlay
//
//  Created by Yuns on 2017/6/4.
//  Copyright © 2017年 yuns. All rights reserved.
//

import UIKit

class ZoneListCell: UICollectionViewCell {

    class func cellWithCollectionView(collectionView : UICollectionView, indexPath: IndexPath) -> ZoneListCell {
        let ID = NSStringFromClass(self)
        let cell : ZoneListCell? = collectionView.dequeueReusableCell(withReuseIdentifier: ID, for: indexPath) as? ZoneListCell
        return cell!
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.contentView.backgroundColor = UIColor.white
        self.addSubViews()
        self.snp_subViews()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    // MARK: - private
    func addSubViews() {
        self.contentView.addSubview(self.titleTextLabel)
        self.contentView.addSubview(self.detailTextLabel)
        self.contentView.addSubview(self.discussImgView)
    }
    
    func snp_subViews() {
        self.titleTextLabel.snp.makeConstraints { (make) in
            make.top.equalTo(self.discussImgView).offset(3)
            make.left.equalTo(self.discussImgView.snp.right).offset(5)
            make.right.equalToSuperview().offset(-10)
        }
        
        self.detailTextLabel.snp.makeConstraints { (make) in
            make.left.right.equalTo(self.titleTextLabel)
            make.bottom.equalTo(self.discussImgView.snp.bottom).offset(-3)
        }
        
        self.discussImgView.snp.makeConstraints { (make) in
            make.width.height.equalTo(50)
            make.left.equalToSuperview().offset(10)
            make.centerY.equalToSuperview()
        }
    }
    
    fileprivate var _listModel : ZoneDiscussItem?
    var listModel : ZoneDiscussItem? {
        set {
            _listModel = newValue
            let imrURL = URL(string: (newValue?.iconUrl)!)
            self.discussImgView.kf.setImage(with: imrURL)
            self.titleTextLabel.text = newValue?.modelName
            self.detailTextLabel.text = newValue?.modelDesc
        }
        
        get {
            return _listModel
        }
    }
    
    // MARK: - setter / getter
    lazy var titleTextLabel : UILabel = {
        let label : UILabel = UILabel()
        label.font = UIFont.systemFont(ofSize: 14)
        label.textColor = RGB(r: 36, g: 36, b: 36)
        return label;
    }()
    
    lazy var detailTextLabel : UILabel = {
        let label : UILabel = UILabel()
        label.font = UIFont.systemFont(ofSize: 12)
        label.textColor = RGB(r: 150, g: 150, b: 150)
        return label;
    }()
    
    lazy var discussImgView : UIImageView = {
        let discussImgView : UIImageView = UIImageView()
        return discussImgView
    }()

}
